### PR TITLE
Net Refactor

### LINKED ISSUE #2433: Builder services not auto-recovering from rt:route:2
Builder services can get into a state where they are not registered with the router. They should be able to auto-recover and get re-registered without having to re-start the router, but currently that is not happening.

### PR DESCRIPTION
**Pull request by reset**
Rewrite Builder's server protocol
* Rewrite the entire network protocol used for inter-server
  communication between Builder services. This is still powered
  by ZeroMQ and Protobufs, it is just much more mature
* Rewrite `net` crate. This crate houses abstractions for defining
  back-end services and all network communication between then.
* Add `http-gateawy` crate. This crate houses abstractions for
  defining gateway services.
* Rewrite large portions of the `builder-protocol` crate. This
  crate houses all information about how connections in the `net`
  and `http-gateway` crate communicate with each other.

tenor-179227442

Closes #2433

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by tashimi** [comment]
tenor-203314641

File: components/builder-depot/src/handlers/integrations.rs [hunk 1]
```diff
@@ -16,34 +16,51 @@ use std::collections::HashMap;
...
     let lock = req.get::<persistent::State<DepotUtil>>().expect(
         "depot not found",
     );
     let depot = lock.read().expect("depot read lock is poisoned");
 
-    match bld_core::integrations::encrypt(&depot.config.key_dir, content) {
-        Ok(c) => Ok(c),
-        Err(_) => Err(status::InternalServerError),
-    }
+    let kp = match BoxKeyPair::get_latest_pair_for(
```
Comment (chefsalim) [inline components/builder-depot/src/handlers/integrations.rs]: This looks like the old code... rebase problem?

Comment (reset) [reply components/builder-depot/src/handlers/integrations.rs]: Yep, got it! Lots of changes here caused a daily rebase of ~30 mins - Please keep an eye out for more of these!

File: components/builder-depot/src/handlers/integrations.rs [hunk 2]
```diff
@@ -13,37 +13,39 @@
...
+use iron::status::{self, Status};
+
 use protocol::originsrv::*;
 use protocol::net::{NetOk, ErrCode};
 use persistent;
-use DepotUtil;
-use bld_core;
+use router::Router;
 
+<<<<<<< Updated upstream
```
Comment (christophermaier) [inline components/builder-depot/src/handlers/integrations.rs]: DANGER DANGER DANGER

File: components/core/src/os/net/mod.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
```
Comment (christophermaier) [inline components/core/src/os/net/mod.rs]: Date

File: components/net/src/app/mod.rs [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            Err(ZError::EINTR) |
+            Err(ZError::ETERM) => return RecvResult::Shutdown,
+            Err(ZError::EFAULT) => panic!("zmq::poll, the provided _items_ was not valid (NULL)"),
+            Err(err) => unreachable!("zmq::poll, returned an unexpected error, {:?}", err),
+        }
+        RecvResult::OnMessage((
+            items[0].is_readable(),
+            items[1].is_readable(),
+            items[2].is_readable(),
+        ))
```
Comment (christophermaier) [inline components/net/src/app/mod.rs]: this is kinda slick

File: components/net/src/app/mod.rs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+use self::config::AppCfg;
+use self::error::{AppError, AppResult};
+use self::dispatcher::{Dispatcher, DispatcherPool};
+use conn::{self, ConnErr};
+use socket::{self, DEFAULT_CONTEXT, ToAddrString};
+
+enum RecvResult {
+    /// Signals to the main loop which sockets have pending messages to be processed. `.0` signals
+    /// that the router socket has messages while `.1` signals the dispatcher queue.
```
Comment (christophermaier) [inline components/net/src/app/mod.rs]: How about `.2`?

File: components/net/src/config.rs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
```
Comment (christophermaier) [inline components/net/src/config.rs]: Date

File: components/net/src/conn/mod.rs [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+use zmq::{self, Error as ZError};
+
+pub use self::error::ConnErr;
+use error::{ErrCode, NetError, NetResult};
+use socket::DEFAULT_CONTEXT;
+
+/// Time to wait before timing out a message receive for a `RouteConn`.
+pub const RECV_TIMEOUT_MS: i32 = 5_000;
+/// Time to wait before timing out a message send for a `RouteBroker` to a router.
+<<<<<<< Updated upstream
```
Comment (christophermaier) [inline components/net/src/conn/mod.rs]: DANGER DANGER DANGER

File: components/net/src/conn/mod.rs [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+    if message.header().has_txn() {
+        send_txn(socket, message)?;
+    }
+    send_body(socket, message)
+}
+
+/// Attempts to wait for a value on this receiver, returning an error if the corresponding
+/// connection has shutdown, or if it waits more than timeout.
+///
+/// This function will always block the current thread if there is no data available and. Once
```
Comment (christophermaier) [inline components/net/src/conn/mod.rs]: "data available and." ?

File: components/builder-originsrv/src/data_store.rs [hunk 9]
```diff
@@ -1235,16 +1235,14 @@ impl DataStore {
     }
 }
 
-fn sync_origins(pool: Pool) -> DbResult<EventOutcome> {
-    error!("I like my butt");
```
Comment (christophermaier) [inline components/builder-originsrv/src/data_store.rs]: tenor-221153222

Comment (robbkidd) [reply components/builder-originsrv/src/data_store.rs]: good bye, sweet quote

File: components/builder-originsrv/src/server/mod.rs [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
```
[the remark this answers is not in the document]
Comment (reset) [reply components/builder-originsrv/src/server/mod.rs]: Actually I've been slowly fixing this. @adamhjk said that we should leave just the initial date and not bother updating each file with a range each year

File: components/builder-originsrv/src/data_store.rs (hunk 9 above)
Comment (adamhjk) [reply components/builder-originsrv/src/data_store.rs]: tenor-3365626

File: components/builder-router/src/server/mod.rs [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        };
+        handler(conn, message, &mut self.servers)
+    }
+
+    fn route_message(&mut self, conn: &SrvConn, message: &mut Message) {
+        let mut x = false;
+        if let Some(txn) = message.txn() {
+            x = txn.is_complete();
+        }
+        // JW TODO: fix this garbage, are you kidding
```
[the remark this answers is not in the document]
Comment (adamhjk) [reply components/builder-router/src/server/mod.rs]: tenor-104400665

File: components/builder-originsrv/src/data_store.rs (hunk 9 above)
Comment (raskchanky) [reply components/builder-originsrv/src/data_store.rs]: tenor-128709179

Comment (chefsalim) [reply components/builder-originsrv/src/data_store.rs]: tenor-160488827

File: components/builder-worker/src/log_forwarder.rs [hunk 12]
```diff
@@ -30,91 +30,77 @@ pub struct LogForwarder {
...
+        &mut self,
+        rz: mpsc::SyncSender<()>,
+        addrs: Vec<(String, String, String)>,
+    ) -> Result<()> {
+        if addrs.len() == 1 {
+            let (_, _, ref log) = addrs[0];
+            println!("Connecting to Job Server Log port, {}", log);
+            self.output_sock.connect(&log)?;
+        } else {
+            warn!("Routing logs to more than one Job Server is not yet implemented");
```
[the remark this answers is not in the document]
Comment (reset) [reply components/builder-worker/src/log_forwarder.rs]: We should definitely see this in the logs - production servers should be configured with the `info` logging level enabled (currently we incorrectly use debug). Our log analysis tools should be set to send alerts during work hours on the presence of `error` messages in logs to all engineers and during non work hours to the engineer on call. We should have a daily digest/dashboard of all errors which also include `warn` messages

File: components/builder-originsrv/src/data_store.rs (hunk 9 above)
Comment (reset) [reply components/builder-originsrv/src/data_store.rs]: tenor-24369314

File: components/builder-worker/src/log_forwarder.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (christophermaier) [reply components/builder-worker/src/log_forwarder.rs]: Sorry, I didn't mean "log" as in "error logging" (which `warn!` will definitely do)... I meant logging in terms of sending job logs to the job server. Previously, if you configured this with multiple job servers (which we explicitly don't support right now) we'd panic. Now there's no connection to the job server, so we're never going to forward anything.

[the remark this answers is not in the document]
Comment (reset) [reply components/builder-worker/src/log_forwarder.rs]: Ah yeah, I'll get it fixed up. We shouldn't panic but we should warn if we have configured it imporperly and do our best. I see that I didn't have it do our best :)

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/276452983).

**Comment by reset** [comment]
This was tested against master already, The Sentinels second build failed due to a travis connectivity error and was unnecessary in the first place

**Comment by thesentinels** [comment]
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/276452983).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/276452983) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

File: components/builder-originsrv/src/data_store.rs (hunk 9 above)
Comment (mgamini) [reply components/builder-originsrv/src/data_store.rs]: tenor-179758398

Comment (christophermaier) [reply components/builder-originsrv/src/data_store.rs]: tenor-202503636

Comment (ryankeairns) [reply components/builder-originsrv/src/data_store.rs]: [](https://media.giphy.com/media/xUA7bilj5YByXZkqnS/giphy.gif)

Comment (ryankeairns) [reply components/builder-originsrv/src/data_store.rs]: tenor-268252881

Comment (ryankeairns) [reply components/builder-originsrv/src/data_store.rs]: tenor-142659970

Comment (reset) [reply components/builder-originsrv/src/data_store.rs]: tenor-254866855

Comment (christophermaier) [reply components/builder-originsrv/src/data_store.rs]: tenor-107020569

**Pull request merged**